pkg/storage/async_store_test.go

```diff
@@ -0,0 +1,212 @@
+			to:   now.Add(time.Duration(i+1) * time.Minute),
+		}))
+
+		chunkRef, err := chunk.ParseExternalKey(chk.UserID, chk.ExternalKey())
+		require.NoError(t, err)
+
+		chunks = append(chunks, chunkRef)
+	}
+
+	return chunks
+}
+
+func buildMockFetchers(num int) []*chunk.Fetcher {
+	var fetchers []*chunk.Fetcher
+	for i := 0; i < num; i++ {
+		fetchers = append(fetchers, &chunk.Fetcher{})
+	}
+
+	return fetchers
+}
+
+func TestAsyncStore_mergeIngesterAndStoreChunks(t *testing.T) {
+	testChunks := buildMockChunkRef(t, 10)
+	fetchers := buildMockFetchers(3)
+	for _, tc := range []struct {
+		name             string
+		storeChunks      [][]chunk.Chunk
+		storeFetcher     []*chunk.Fetcher
+		ingesterChunkIDs []string
+		ingesterFetcher  *chunk.Fetcher
+		expectedChunks   [][]chunk.Chunk
+		expectedFetchers []*chunk.Fetcher
+	}{
+		{
+			name: "no chunks from both",
+		},
+		{
+			name:             "no chunks from ingester",
+			storeChunks:      [][]chunk.Chunk{testChunks},
+			storeFetcher:     fetchers[0:1],
+			expectedChunks:   [][]chunk.Chunk{testChunks},
+			expectedFetchers: fetchers[0:1],
+		},
+		{
+			name:             "no chunks from querier",
+			ingesterChunkIDs: convertChunksToChunkIDs(testChunks),
+			ingesterFetcher:  fetchers[0],
+			expectedChunks:   [][]chunk.Chunk{testChunks},
+			expectedFetchers: fetchers[0:1],
+		},
+		{
+			name: "nothing common",
+			storeChunks: [][]chunk.Chunk{
+				testChunks[0:5],
+			},
+			storeFetcher:     fetchers[0:1],
+			ingesterChunkIDs: convertChunksToChunkIDs(testChunks[5:]),
+			ingesterFetcher:  fetchers[1],
+			expectedChunks: [][]chunk.Chunk{
+				testChunks[0:5],
+				testChunks[5:],
+			},
+			expectedFetchers: fetchers[0:2],
+		},
+		{
+			name: "common chunks, different fetchers",
+			storeChunks: [][]chunk.Chunk{
+				testChunks[0:5],
+				testChunks[5:],
+			},
+			storeFetcher:     fetchers[0:2],
+			ingesterChunkIDs: convertChunksToChunkIDs(testChunks[5:]),
+			ingesterFetcher:  fetchers[2],
+			expectedChunks: [][]chunk.Chunk{
+				testChunks[0:5],
+				testChunks[5:],
+			},
+			expectedFetchers: fetchers[0:2],
+		},
+		{
+			name: "different chunks, common fetchers",
+			storeChunks: [][]chunk.Chunk{
+				testChunks[0:2],
+				testChunks[2:5],
+			},
+			storeFetcher:     fetchers[0:2],
+			ingesterChunkIDs: convertChunksToChunkIDs(testChunks[5:]),
+			ingesterFetcher:  fetchers[1],
+			expectedChunks: [][]chunk.Chunk{
+				testChunks[0:2],
+				testChunks[2:],
+			},
+			expectedFetchers: fetchers[0:2],
+		},
+		{
+			name: "different chunks, different fetchers",
+			storeChunks: [][]chunk.Chunk{
+				testChunks[0:2],
+				testChunks[2:5],
+			},
+			storeFetcher:     fetchers[0:2],
+			ingesterChunkIDs: convertChunksToChunkIDs(testChunks[5:]),
+			ingesterFetcher:  fetchers[2],
+			expectedChunks: [][]chunk.Chunk{
+				testChunks[0:2],
+				testChunks[2:5],
+				testChunks[5:],
+			},
+			expectedFetchers: fetchers[0:3],
+		},
+		{
+			name: "duplicate chunks from ingesters",
```

Comment:
I think a test case covering duplicate chunks from ingester and queriers would be good too.

Reply:
The [common chunks, different fetchers](https://github.com/grafana/loki/pull/2601/files/3ac545040fcd4fd6cc9b0bb2c0245b058cdb2e57#diff-ce5f96fb74a428c2a4158501a3fd86ffR123-R129) test covers it I guess or am I missing another case here?

Reply:
ah i think it's a bit confusing to swap between "duplicate" and "common", but you are right that case is what i'm looking for. I think i would suggest using "duplicate" instead of common in those test names

Reply:
Sounds good. Changed the test names.